pkg/datapath/linux/ipsec/ipsec_linux_test.go

```diff
@@ -0,0 +1,150 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build privileged_tests
+
+package ipsec
+
+import (
+	"net"
+	"os"
+	"testing"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/vishvananda/netlink"
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) { TestingT(t) }
+
+type IPSecSuitePrivileged struct{}
+
+var _ = Suite(&IPSecSuitePrivileged{})
+
+var (
+	path             = "ipsec_keys_test"
+	dfltKeysDat    = []byte("hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF\n")
+	scopedKeysDat  = []byte("hmac(sha256) abcdefghijklmnopqrstuvwzyzABCDEF cbc(aes) abcdefghijklmnopqrstuvwzyzABCDEF foobar\n")
+	invalidKeysDat = []byte("test abcdefghijklmnopqrstuvwzyzABCDEF test abcdefghijklmnopqrstuvwzyzABCDEF\n")
+)
+
+func (p *IPSecSuitePrivileged) TestLoadKeysNoFile(c *C) {
+	err := LoadIPSecKeysFile(path)
+	c.Assert(err, ErrorMatches, "failed to load IPSec Keys File ipsec_keys_test: open ipsec_keys_test: no such file or directory")
+}
+
+func (p *IPSecSuitePrivileged) TestInvalidLoadKesyFile(c *C) {
+	f, err := os.Create(path)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(invalidKeysDat)
+	c.Assert(err, IsNil)
+	f.Sync()
+	err = LoadIPSecKeysFile(path)
+	c.Assert(err, IsNil)
+	err = os.Remove(path)
+	c.Assert(err, IsNil)
+
+	_, local, err := net.ParseCIDR("1.1.3.4/16")
+	c.Assert(err, IsNil)
+	_, remote, err := net.ParseCIDR("1.2.3.4/16")
+	c.Assert(err, IsNil)
+	err = UpsertIPSecEndpoint(local, remote)
+	c.Assert(err, NotNil)
+}
+
+func (p *IPSecSuitePrivileged) TestLoadKesyFile(c *C) {
+	path := "ipsec_keys_test"
+	f, err := os.Create(path)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(dfltKeysDat)
+	c.Assert(err, IsNil)
+	_, err = f.Write(scopedKeysDat)
+	c.Assert(err, IsNil)
+	f.Sync()
+	err = LoadIPSecKeysFile(path)
+	c.Assert(err, IsNil)
+	err = os.Remove(path)
+	c.Assert(err, IsNil)
+}
+
+func (p *IPSecSuitePrivileged) TestUpsertIPSecEquals(c *C) {
+	_, local, err := net.ParseCIDR("1.2.3.4/16")
+	c.Assert(err, IsNil)
+	_, remote, err := net.ParseCIDR("1.2.3.4/16")
+	c.Assert(err, IsNil)
+
+	key := &ipSecKey{
+		Spi:   1,
+		ReqID: 1,
+		Auth:  &netlink.XfrmStateAlgo{Name: "hmac(sha256)", Key: []byte("abcdefghijklmnopqrstuvwzyzABCDEF")},
+		Crypt: &netlink.XfrmStateAlgo{Name: "cbc(aes)", Key: []byte("abcdefghijklmnopqrstuvwzyzABCDEF")},
+	}
+
+	ipSecKeysGlobal["1.2.3.4"] = key
+	ipSecKeysGlobal[""] = key
+
+	err = UpsertIPSecEndpoint(local, remote)
+	c.Assert(err, IsNil)
+
+	err = DeleteIPSecEndpoint(remote.IP, local.IP)
```

Comment:
Is there any good test you could put in between the `Upsert()`/`Delete()` that checks for the key or something that will be used from the key, ideally without testing directly the implementation?

Reply:
We could do a lookup from the netlink side. But I haven't implemented the helper to do that yet. Having a ``FindIPSecEndpoint` helper would be great but I've not actually had any reason to use it except here.